keysync_initiator.go

```diff
@@ -0,0 +1,207 @@
+// We are using those fields as follows:
+//
+// When the requesting enclave sends a request to the remote enclave, it sets
+// the following fields in the attestation document:
+//
+// Attestation document(
+//   Nonce:       Remote enclave's nonce
+//   User data:   Requesting enclave's nonce
+//   Public key:  Requesting enclave's NaCl box public key
+// )
+//
+// The remote enclave then generates its own attestation document containing
+// the following fields:
+//
+// Attestation document(
+//   Nonce:       The nonce the requester provided in its attestation document
+//   User data:   Encrypted key material
+//   Public key:
+// )
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/hf/nitrite"
+	"golang.org/x/crypto/nacl/box"
+)
+
+// RequestKeys asks a remote enclave to share its key material with us, which
+// is then written to the provided variable.
+//
+// This is only necessary if you intend to scale enclaves horizontally.  If
+// you will only ever run a single enclave, ignore this function.
+func RequestKeys(addr string, keyMaterial any) error {
+	errStr := "failed to request key material"
+
+	// First, request a nonce from the remote enclave.
+	theirNonce, err := requestNonce(addr)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now, create our own nonce.
+	ourNonce, err := newNonce()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Next, create a key that the remote enclave is going to use to encrypt
+	// its key material.
+	boxKey, err := newBoxKey()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now create an attestation document containing our nonce, the remote
+	// enclave's nonce, and the key material that they remote enclave is
+	// supposed to use.
+	ourAttDoc, err := attest(theirNonce[:], ourNonce[:], boxKey.pubKey[:])
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Send our attestation document to the remote enclave, and get theirs in
+	// return.
+	theirAttDoc, err := requestAttDoc(addr, ourAttDoc)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Finally, verify the attestation document and extract the key material.
+	if err := processAttDoc(theirAttDoc, &ourNonce, boxKey, keyMaterial); err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	return nil
+}
+
+// requestNonce requests a nonce from the remote enclave specified by 'addr'.
+func requestNonce(addr string) (nonce, error) {
+	errStr := "failed to fetch nonce from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathNonce)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	maxReadLen := base64.StdEncoding.EncodedLen(nonceLen)
+	body, err := io.ReadAll(newLimitReader(resp.Body, maxReadLen+1))
```

Comment:
Is this `+1` still needed after 5cd952ee?

Reply:
No, good catch. Fixed in 58e8bf9dc99f63a25ed624d212342dc2c0b3285c.